arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -469,72 +463,33 @@ public fun <A> Resource<A>.asFlow(): Flow<A> =
  * This API is useful for building inter-op APIs between [Resource] and non-suspending code, such as Java libraries.
  */
 @DelicateCoroutinesApi
-public suspend fun <A> Resource<A>.allocated(): Pair<A, suspend (ExitCase) -> Unit> {
-  val effect = ResourceScopeImpl()
-  val allocated: A = invoke(effect)
-  val release: suspend (ExitCase) -> Unit = { e ->
-    val suppressed: Throwable? = effect.cancelAll(e)
-    val original: Throwable? = when(e) {
-      ExitCase.Completed -> null
-      is ExitCase.Cancelled -> e.exception
-      is ExitCase.Failure -> e.failure
-    }
-    original?.apply {
-      suppressed?.let(::addSuppressed)
-    }?.let { throw it }
-  }
-  return Pair(allocated, release)
+public suspend fun <A> Resource<A>.allocated(): Pair<A, suspend (ExitCase) -> Unit> = with(ResourceScopeImpl()) {
+  bind() to this::cancelAll
 }
 
-@JvmInline
-private value class ResourceScopeImpl(
-  private val finalizers: Atomic<List<suspend (ExitCase) -> Unit>> = Atomic(emptyList()),
-) : ResourceScope {
-  override suspend fun <A> Resource<A>.bind(): A = invoke(this@ResourceScopeImpl)
-  
-  override suspend fun <A> install(acquire: suspend AcquireStep.() -> A, release: suspend (A, ExitCase) -> Unit): A =
-    bracketCase({
-      val a = acquire(AcquireStep)
-      val finalizer: suspend (ExitCase) -> Unit = { ex: ExitCase -> release(a, ex) }
-      finalizers.update(finalizer::prependTo)
-      a
-    }, ::identity, { a, ex ->
-      // Only if ExitCase.Failure, or ExitCase.Cancelled during acquire we cancel
-      // Otherwise we've saved the finalizer, and it will be called from somewhere else.
-      if (ex != ExitCase.Completed) {
-        val e = cancelAll(ex)
-        val e2 = kotlin.runCatching { release(a, ex) }.exceptionOrNull()
-        e?.apply {
-          e2?.let(::addSuppressed)
-        }?.let { throw it }
-      }
-    })
+@PublishedApi
+internal class ResourceScopeImpl : ResourceScope {
```

Comment:
Made it not a value class to prevent API surface explosion, but it can be undone (especially if we use the #3515 technique)

Reply:
Could also move the implementing of the ResourceScope to an `val scope = object : ResourceScope {...}` - effectively making this class have `val scope: ResourceScope` and `fun cancelAll`, reducing the exposed ABI

Reply:
Not sure how the `cancelAll` would be exposed though? Can you clarify a little more?

Reply:
I found a way to hide it completely. Not sure if making it a `value class` is a good idea though, since it'll always be boxed.

Reply:
No benefit to the value class when it's being treated as a ResourceScope as it'll be boxed. I'm not sure how the `::cancelAll` reference would get represented in the byte-code, but likely it's boxed too. Simpler to reason over a simple class and instance method reference.